README.md
# shiny-sampling-grid-generator

Requires the following packages:
'ggplot2','summarytools','dplyr', 'shinydashboard', 'leaflet', 'shinyjs'

and a subfolder /www
which should contain file releases.md


releases.md
### About this app
<br>
<p style=" font-size:16px; text-align:justify;">
This Shiny application provide an easy tool for designing a sampling grid for field inventories
and for visual remote sensing data analysis. You can use this tool to create a sampling grid also for 
<a href="https://openforis.org/tools/collect-earth/" target="_blank">Open Foris Collect Earth</a>. 
<br>
The sampling grid can be delineated with the help of an imported (polygon or multipolygon) shapefile, or just by
giving in the coordinate reference system (CRS) and extent of the area of interest (AOI).
<br><br>
The application can create a systematic network of points and these are called as ‘clusters’. Each cluster
can contain 1-x sampling points, and these are called as ‘plots’. The user should define the AOI, distances
between clusters in X (West-East) and Y (South-North) directions. A hexagonal grid is also available.
<br><br>
<a href="https://drive.google.com/file/d/19oxIFQFLA0dCjxM4zBkJbSpUR1IiHYeE/view" target="_blank">User's Guide is here</a> (PDF)

<br><br>

### Releases
<p style=" font-size:14px; text-align:justify;">
17 August 2022&nbsp;&nbsp;
<br>
Lauri Vesa and Stefano Giaccio, FAO
<br>
<b>version:</b> 1.3


server.R
##########################################################################################################
# Plot grid generator
# (c) Lauri Vesa, Stefano Giaggio, FAO Forestry Department, Rome
#
# see
# https://cran.rstudio.com/web/packages/rgdal/vignettes/CRS_projections_transformations.html
#
# Last update:
# 24.2.2022 LV, all spatial object to sf (from sp)
# 28.1.2022 LV, taken out several loaded packages
# 30.9.2020 LV, replaced mapview package with leafem,  https://www.rdocumentation.org/packages/mapview/versions/2.7.8/topics/addMouseCoordinates 
#
#
##########################################################################################################
options("rgdal_show_exportToProj4_warnings"="none")
options(stringsAsFactors = FALSE)
options(digits = 15)

library('stringr')
library('sf')
library('tidyr')    
# library('raster')    # extent-function is in this package
library('leaflet')
library('dplyr')
library('sfheaders') # https://stackoverflow.com/questions/61215968/creating-sf-polygons-from-a-dataframe
library('leafem')
library('geosphere')
library('nngeo')


sf::sf_use_s2(FALSE)

my_zip_tools <- Sys.getenv("R_ZIPCMD", "zip")


try(source("Grid-Generate_clusters.R"))
try(source("Grid-Generate_plots.R"))
try(source("Grid-Generate_plot_polygons.R"))

writeOutputFiles <- function(clusters, plot_points, plot_polygons) {   
  prevWD <- getwd()
  tmpdir <- tempdir()
  setwd(tempdir())
  
  if (!file.exists(".//maps")){
    dir.create(file.path('.', 'maps'), showWarnings = FALSE)
  }
  
  clusters       <- clusters %>% dplyr::select(Cluster_ID,	RowNo, ColNo, zone,	X_map, Y_map, X, Y)
  clusters_df    <- st_drop_geometry(clusters)
  plot_points_df <- st_drop_geometry(plot_points)
  
  
  # cluster point data
  st_write(clusters_df, "cluster_point.csv", delete_dsn=TRUE)
  st_write(clusters, dsn="maps", layer="cluster_point", driver="ESRI Shapefile", overwrite_layer = TRUE, append=FALSE )

  # plot point data
  st_write(plot_points_df, "plot_points.csv", delete_dsn=TRUE)
  # 
  # # create plot input data for Collect Earth
  if (nrow(plot_points_df) == nrow(clusters_df)) {
     id <- c(plot_points_df$Cluster_ID)
   } else {
     id <- c(paste( plot_points_df$Cluster_ID, "_", plot_points_df$Plot_ID, sep="" ))
  }
  ceDF <- data.frame(id, stringsAsFactors = FALSE)
  ceDF$YCoordinate <- plot_points_df$Y
  ceDF$XCoordinate <- plot_points_df$X
  ceDF$elevation <- 0; ceDF$aspect <- 0; ceDF$slope <- 0
  st_write(ceDF, "plot_points_CE.csv", delete_dsn=TRUE)
  
  # plot point data, shape
  st_write(plot_points, dsn="maps", layer="plot_point", driver="ESRI Shapefile", overwrite_layer = TRUE, append=FALSE)
  # # plot polygon data, shape
  st_write(plot_polygons, dsn="maps", layer="plot_polygons", driver="ESRI Shapefile", overwrite_layer = TRUE, append=FALSE)
  
  setwd(prevWD)
} 

# see https://www.rdocumentation.org/packages/vetools/versions/1.3-28/topics/get.shape.range
get.shape.range <- function (shape) 
{
  
  SHP.range = matrix(ncol = 4, nrow = length(shape))
  for (i in 1:length(shape)) {
    d = slot(shape, "polygons")[[i]]
    SHP.sub = matrix(ncol = 4, nrow = length(slot(d, "Polygons")))
    for (j in 1:length(slot(d, "Polygons"))) {
      d.sub = slot(d, "Polygons")[[j]]
      d.sub = slot(d.sub, "coords")
      SHP.sub[j, 1:2] = range(d.sub[, 1])
      SHP.sub[j, 3:4] = range(d.sub[, 2])
    }
    d = matrix(apply(SHP.sub, 2, range), ncol = 4)
    SHP.range[i, 1:2] = diag(d[1:2, 1:2])
    SHP.range[i, 3:4] = diag(d[1:2, 3:4])
  }
  d = matrix(apply(SHP.range, 2, range), ncol = 2)
  d = matrix(apply(d, 2, range), ncol = 4)
  colnames(d) <- c("Long.start", "Long.end", "Lat.start", 
                   "Lat.end")
  return(d)
}


enableBookmarking("url")
options(shiny.maxRequestSize=35*1024^2) 
map_projection = 0

getmode <- function(v) {
  uniqv <- unique(v)
  uniqv[which.max(tabulate(match(v, uniqv)))]
}

target_distance <- function(shape,no_target_clusters, input_hexa_equaldist){
  if (no_target_clusters > 2 & !is.null(shape)) {
    
    target_grid <- st_sample(shape, size=no_target_clusters, type='regular')

    target_dist <- getmode(diff(st_coordinates(target_grid)[,1]))
    rm(target_grid)
    if (input_hexa_equaldist) {target_dist = target_dist * 1/sqrt(sqrt(3/4)) }
      
    return(round(target_dist,2))
  } else {
    return(0)
  }
}


server <- function(input, output, session) {

    shape <- reactive({
    if (!is.null(input$shpFile)){
      shpDF <- input$shpFile
      if (nrow(shpDF) < 4) {
        cat(" Not all 4 files read /n")
        shpDF <- NULL
        return()  
      }
      
       prevWD <- getwd()
       uploadDirectory <- dirname(shpDF$datapath[1])
       setwd(uploadDirectory)
       for (i in 1:nrow(shpDF)){
         file.rename(shpDF$datapath[i], shpDF$name[i])
       }
       shpName <- shpDF$name[grep(x=shpDF$name, pattern="*.shp")]
       fName   <- substr(shpName,1, nchar(shpName)-4)
       if (is.vector(fName)) {
         shpName <- as.character( shpName[1] )
         fName   <- as.character( fName[1] )
       } else {
         return()
       }
#       print(str(fName))
       shpPath <- paste(uploadDirectory, shpName, sep="/")

       setwd(prevWD)
       shpFile <- sf::st_read(dsn=shpPath, layer= fName)

      
       # antimeridian case processing
       # https://stackoverflow.com/questions/55162548/is-there-a-better-way-for-handling-spatialpolygons-that-cross-the-antimeridian
       # shape <- shape %>%
       #   st_wrap_dateline(options = c("WRAPDATELINE=YES"), quiet = TRUE) %>%
       #   st_union()
       
       print("Shape imported")
       output$map_proj1 <- renderPrint({ st_crs(shpFile)$proj4string })
       return(shpFile)
    } else {
        return()
    }
  })
  
  shape4326 <- reactive({
    shape = shape()
    if (is.null(shape)) return(NULL)
    
    print("map_projection missing?")
 #   print(st_crs(shape))

    map_projection <- as.integer( sf::st_crs(shape)$epsg )
#    print(map_projection)
    if (is.na(map_projection)) {
      output$map_proj1 <- renderPrint({ "Missing or unsupported CRS" })
      return(NULL)
    }
    
    if (map_projection != 4326) { 
        shape = st_transform( shape, 4326 ) 
        updateNumericInput(session, "cluster_x_distance", value=100000)
        updateNumericInput(session, "cluster_y_distance", value=100000)
      } else {
        updateNumericInput(session, "cluster_x_distance", value=10)
        updateNumericInput(session, "cluster_y_distance", value=10)
      }
    
    
    return( shape )
  })
  
  # Insert and delete plots  
  inserted_div_ids <- c()
  inserted_ids <- c()
  
  observeEvent(input$insertPlotBtn, {
    id <- input$insertPlotBtn # incremental number
    insertPlotRow(id, 0, 0)
  }, ignoreInit = TRUE)
  

  output$fileUploaded <- reactive({
     if (!is.null(shape())) {
       shape = shape()
       return(!is.null(shape))
     } else {
       return(FALSE)
     }

  })
  outputOptions(output, "fileUploaded", suspendWhenHidden = FALSE)

  
  # https://stackoverflow.com/questions/49186866/using-checkbox-to-enable-disable-output
  react <- reactive({
    shape = shape()
    (input$use_map_extent==TRUE | input$use_clipping_map==TRUE) & !is.null(shape)
  })

   
  observeEvent(react(), {
    
    if (react()) {
      shinyjs::hide("minx")
      shinyjs::hide("maxx")
      shinyjs::hide("miny")
      shinyjs::hide("maxy")
    } else {
      shinyjs::show("minx")
      shinyjs::show("maxx")
      shinyjs::show("miny")
      shinyjs::show("maxy")
    }
  })


  observeEvent(shape4326(), {
    leafletProxy("plotmap") %>% clearShapes()
    # Clear the map elements
    leafletProxy("plotmap") %>% clearGroup("Cluster points")
    leafletProxy("plotmap") %>% clearGroup("Plot points")
    leafletProxy("plotmap") %>% clearGroup("Plot shapes")

    map = shape4326()

    leafletMap = leafletProxy("plotmap", data = map) %>%
      addPolygons(
        fillColor = "transparent",
        weight = 1,
        opacity = 0.7
      )
    
    # https://stackoverflow.com/questions/51392181/setting-boundaries-of-leaflet-with-sf-geometry-in-r
    e <- st_bbox(map) %>% as.vector()
    
    leafletMap %>% 
      addTiles() %>% 
      fitBounds(e[1], e[2], e[3], e[4])

  }, ignoreInit = TRUE)

  
  observeEvent(input$target_dist_button, {
      input_hexa_equaldist = FALSE
      if (input$cluster_hexagonal & input$cluster_hexagonal_dist) input_hexa_equaldist = TRUE 
      outText1 <- target_distance(shape(), input$target_clusters, input_hexa_equaldist)
      updateTextInput(session, "target_distance", value = outText1)
  })
  
  
    nDistance <- reactive( {
    new <- input$target_distance
    new <- ifelse(is.na(new) | is.null(new) | trimws(new)=="", "", new )
    new
  })
  
  observeEvent(input$applydistance, {
    if (nDistance() != "") {
      estim_distance <- nDistance()
      print(estim_distance)
      updateNumericInput(session, "cluster_x_distance", value=estim_distance)
      updateNumericInput(session, "cluster_y_distance", value=estim_distance)
    }
  })  
  

  observe( {
    if (input$cluster_hexagonal) {
      shinyjs::show("cluster_hexagonal_dist")
      shinyjs::toggleState("cluster_hexagonal_dist", input$cluster_hexagonal)
    } else {
      shinyjs::hide("cluster_hexagonal_dist")  
    }
    })

  observe( {
    if (input$cluster_hexagonal & input$cluster_hexagonal_dist) {
      shinyjs::disable("cluster_y_distance")
    } else {
      shinyjs::enable("cluster_y_distance")  
    }
  })
  
#  observe( {
#    if (input$drop_rowcol) {
#      shinyjs::show("dropfactor")
#    } else {
#      shinyjs::hide("dropfactor")
#    }
#  })
  
  
    

  observeEvent(input$run, {
    if (!is.null(shape())) map <- shape()
    
    if (exists('map') & (input$use_map_extent | input$use_clipping_map)) {
      e     <- st_bbox(map)
      min_x <- e$xmin
      min_y <- e$ymin
      max_x <- e$xmax
      max_y <- e$ymax
      rm(e)
      map_projection <- as.integer( st_crs(map)$epsg )
    } else {
      min_x <- input$minx
      min_y <- input$miny
      max_x <- input$maxx
      max_y <- input$maxy
      map_projection <- as.integer(substring(trimws(input$projection), 6,))
    }
    
    # the next line will show CRS also if no shape file is loaded
    output$map_proj1 <- renderPrint({ st_crs(map_projection)$proj4string })

    # Too many arguments - will refactor
    # clusters is a sf object
    print(map_projection)
    print("###")
    
    clusters <- generate_clusters(
      cluster_start_random = input$cluster_start_random,
      cluster_hexagonal = input$cluster_hexagonal,
      cluster_hexagonal_dist = input$cluster_hexagonal_dist,
      map_projection,
      min_x = min_x,
      max_x = max_x,
      min_y = min_y,
      max_y = max_y,
      cluster_x = input$cluster_x,
      cluster_y = input$cluster_y,
      cluster_x_distance = input$cluster_x_distance,
      cluster_y_distance = input$cluster_y_distance,
      first_cluster_id = input$first_cluster_id
    )
    
    ## Clipping by using shape ----

    if ( input$use_clipping_map & exists('map')) {
      tryCatch({
#        sf::sf_use_s2(FALSE)  # https://stackoverflow.com/questions/68478179/how-to-resolve-spherical-geometry-failures-when-joining-spatial-data

        if (map_projection != 4326) { 
            map_clip <- st_transform(map, 4326) 
        } else { 
            map_clip <- map
        }
        # https://community.rstudio.com/t/shinyapps-working-locally-but-failing-on-build-related-to-sf/78245/3
        map_clip <- sf::st_buffer(map_clip,0)
        # do clipping
        print("DOING CLIPPING")

        clusters <- st_intersection(clusters, map_clip)
        rm(map_clip)
        },
        # but if an error occurs
        error = function(error_message) {
          print("Failed st_itersection")
          return(NA)
        }
      )
      # renumber clusters
      clusters$Cluster_ID <- 1:nrow(clusters) + input$first_cluster_id - 1
    }
    
    clusters$Cluster_ID <- as.integer(clusters$Cluster_ID)

    #    if (is.na(clusters)) {break}
    print("END CLUSTER GENERATING")
    
    # Clear the map
    leafletProxy("plotmap") %>% clearGroup("Cluster points")
    leafletProxy("plotmap") %>% clearGroup("Plot points")
    leafletProxy("plotmap") %>% clearGroup("Plot shapes")
    # Add cluster point markers
    
    print("MAIN 1")

    leafletProxy("plotmap", data = st_drop_geometry(clusters)) %>%
      addCircleMarkers(~X,~Y,
        radius = 4,
        stroke = FALSE,
        fillOpacity = 0.5,
        group = "Cluster points"
      )
    
    print("MAIN 2")
    # Generate plot points
    # Create three separate (ids, x, and y) vectors from plot coordinates to feed the generate_plots function
    ids <- c()
    xs  <- c()
    ys  <- c()
    # xy_swap <- c()
    
    # sum function is used to eliminate NA cases
    for (i in inserted_ids) {
      ids     <- c(ids, sum(0, as.numeric(input[[paste0("plot_id_input_", i)]]), na.rm=T) )
      xs      <- c(xs,  sum(0, as.numeric(input[[paste0("plot_x_input_",  i)]]), na.rm=T) )
      ys      <- c(ys,  sum(0, as.numeric(input[[paste0("plot_y_input_",  i)]]), na.rm=T) )
      # if ( input$plot_type=="R" & (input$plot_dim_1 != input$plot_dim_2)) {
      #   xy_swap <- c(xy_swap, input[[paste0("plot_xy_input_",  i)]])
      # } else {
      #   xy_swap <- c(xy_swap, FALSE)
      # }
    }
    
    
    if (length(inserted_ids)==0) {
      ids <- c(1)
      xs  <- c(0)
      ys  <- c(0)
#      xy_swap <- c(FALSE)
    }
    
    plot_list <- data.frame(ids, xs, ys)
    names(plot_list) <- c("No", "X", "Y") 
    # plot_list <- data.frame(ids, xs, ys, xy_swap)
    # names(plot_list) <- c("No", "X", "Y", "xy_swap") 
    
    
    # plot_points is a sf object
    plot_points <- generate_plots(
      clusters = clusters,
      Plot_type=input$plot_type,
      plot_list,
      map_projection
    )

    if (exists("clusters")) {
      output$number_clusters <- renderText({ paste0("Number of clusters: ", nrow(clusters)) }) }

    if (exists("plot_points")) {
      output$number_plots    <- renderText({ paste0("Number of plots: ", nrow(plot_points)) }) }

    print(str(plot_points))
    plot(plot_points)
    
    # Add plot point markers
    leafletProxy("plotmap", data = st_drop_geometry(plot_points)) %>%
        addCircleMarkers(~X,~Y,
          radius = 2,
          color = "#ff0000",
          stroke = FALSE,
          fillOpacity = 0.5,
          group = "Plot points"
        )
    
  
    # Generate plot polygons

    print("MAIN 3")
    plot_polygons <- generate_plot_polygons(
      p = plot_points,
      Plot_type =input$plot_type,
      PlotCorner_X =input$plot_corner_x,
      PlotCorner_Y =input$plot_corner_y,
      Plot_ref_point =input$plot_reference_point,
      Plot_dim1 =input$plot_dim_1,
      Plot_dim2 =input$plot_dim_2,
      map_projection
    )

    writeOutputFiles(clusters, plot_points, plot_polygons)
    
    
    output$downloadDataCSV <- downloadHandler(
      filename = "grid_point_csv.zip",
      content = function(fname) {
        prevWD <- getwd()
        setwd(tempdir())
        file_list <- list.files(pattern = "\\.csv$", full.names = FALSE, ignore.case=TRUE)
        zip( 
          zipfile=fname, 
          files=file_list, 
          zip=my_zip_tools
        )
        setwd(prevWD)
      },
      contentType = "application/zip"
    ) 

    output$downloadDataShape <- downloadHandler(
      filename = "grid_shapes.zip",
      content = function(fname) {
        prevWD <- getwd()
        setwd(tempdir())
        try(setwd("maps"), silent=T)
        file_list <- list.files( full.names = FALSE)
        zip(
          zipfile=fname,
          files=file_list,
          zip=my_zip_tools
        )
        setwd(prevWD)
      },
      contentType = "application/zip"
    )
  
    # Add plot polygon markers
    leafletProxy("plotmap", data = plot_polygons) %>%
      addPolygons(
        fillColor = "transparent",
        weight = 1,
        group = "Plot shapes"
      )
  }, ignoreInit = TRUE);

  
  output$map_proj1 <- renderPrint({ 
    if (map_projection!=0) st_crs(map_projection)$proj4string })
  
  # Add map
  output$plotmap <- renderLeaflet({
    leaflet() %>%
      addTiles(group = "OSM (default)") %>%
      addProviderTiles(providers$Stamen.TonerLite, group = "Toner Lite") %>%
      addLayersControl(
        baseGroups = c("OSM (default)", "Toner Lite"),
        overlayGroups = c("Cluster points", "Plot points", "Plot shapes"),
        options = layersControlOptions(collapsed = FALSE)) %>%
      hideGroup("Plot points") %>%
      hideGroup("Plot shapes")  %>%
      leafem::addMouseCoordinates()
  })

  onRestore(function(state) {
    # Get the id field names (plot_id_input_1, plot_id_input_2, ...)
    plot_id_field_names <- names(state$input)[grep("plot_id_input", names(state$input))]
    m   <- regexpr('[0-9]+$', plot_id_field_names)
    ids <- regmatches(plot_id_field_names, m)
    # As the id increases when inserting new coords from the UI, we restore the order here
    ids <- sort(ids)
    
    for (i in 1:length(ids)) {
      x = state$input[[paste0("plot_x_input_", ids[i])]]
      y = state$input[[paste0("plot_y_input_", ids[i])]]
      insertPlotRow(ids[i], x, y)
    }
  })
  
  insertPlotRow <- function(plot_id, x, y) {
    div_id <- paste0("plot_div_", plot_id) # wrap in div to ease removal
    plot_id_input_id <- paste0("plot_id_input_", plot_id)
    plot_x_input_id  <- paste0("plot_x_input_", plot_id)
    plot_y_input_id  <- paste0("plot_y_input_", plot_id)
#    plot_xy_input_id <- paste0("plot_xy_input_", plot_id)
    delete_button_id <- paste0("delete_plot_", plot_id)
    
    # Insert id, x and y fields, with related values
    insertUI(
      selector = "#plots",
      ui = tags$div(
        fluidRow(
          # column(width = 2, textInput(plot_id_input_id, "Plot ID", plot_id)),
          # column(width = 3, numericInput(plot_x_input_id, "X", x)),
          # column(width = 3, numericInput(plot_y_input_id, "Y", y)),
          # column(width = 1, actionButton(delete_button_id, "Delete this plot"))
          column(width = 2, textInput(plot_id_input_id, "Plot ID", plot_id)),
          column(width = 3, numericInput(plot_x_input_id, "X", x)),
          column(width = 3, numericInput(plot_y_input_id, "Y", y)),
          # if ( input$plot_type=="R" & (input$plot_dim_1 != input$plot_dim_2)) {
          #   column(width = 3, selectInput("plot_xy_input_id", "Rotate plot",
          #               c("None" = "0",
          #                 "Counter-clockwise (-90)" = "-90",
          #                 "Clockwise (90)" = "90")))},
#            column(width = 3, checkboxInput(plot_xy_input_id, "Swap dimensions", FALSE))},
          column(width = 1, actionButton(delete_button_id, "Delete plot"))
        ),
        hr(),
        id = div_id
      )
    )
    
    # Delete button
    observeEvent(input[[delete_button_id]], {
      pos <- match(c(plot_id), inserted_ids)
      inserted_div_ids <<- inserted_div_ids[-pos]
      inserted_ids <<- inserted_ids[-pos]
      
      removeUI(selector = paste0("#", div_id))
    },
    ignoreInit = TRUE)
    
    inserted_div_ids <<- c(div_id, inserted_div_ids)
    inserted_ids <<- c(plot_id, inserted_ids)
  }

  # readShape2 <- function(shapefile_path, layer) {
  #   tryCatch(
  #     shapes <- st_read(dsn = shapefile_path, layer = layer),
  #     error = function(error_message) {
  #       message("Shape not could not be read")
  #       return(NULL)
  #     }
  #   )
  #   return(shapes)
  # }
  
}


Grid-Generate_plots.R
## +++++++++++++++++++++++++++++++++++++++++++++++
## Generates plot locations (X, Y) ------
## p = plot point dataframe

# https://stackoverflow.com/questions/5173692/how-to-return-number-of-decimal-places-in-r
decimalplaces <- function(x) {
  ifelse(abs(x - round(x)) > .Machine$double.eps^0.5,
         nchar(sub('^\\d+\\.', '', sub('0+$', '', as.character(x)))),
         0)
}

########################################################

generate_plots <- function(clusters,
                           Plot_type,
                           plot_list,
                           map_projection) {
  
  DT_Clusters <- st_drop_geometry(clusters)
  
  deci_x <- max(decimalplaces(DT_Clusters$X_map)) 
  deci_y <- max(decimalplaces(DT_Clusters$Y_map)) 
  
  
  nclus       <- nrow(DT_Clusters)
  nplots      <- nrow(plot_list)

  if (Plot_type=='NFMA') {
    nplots <- 4
    rm(plot_list)
    plot_list <- data.frame( No=c(1:4), X=c(250,250,750,750), Y=c(250,750,750,250) )
  }

  # for 4326, compute bearing (degrees) and distance (m)
  if (map_projection == 4326) {
    for (i in (1:nplots)) {
      plot_list$azimuth[i]  <- nngeo::st_azimuth(st_point(c(0, 0)), st_point(c(plot_list$X[i], plot_list$Y[i])))
    }
    plot_list$distance <- sqrt((plot_list$X^2 + plot_list$Y^2))
    print(str(plot_list))
  }
  
  print("PLOT 1")
  p <- as.data.frame(matrix(0, ncol = 5, nrow = nclus * nplots))
  colnames(p)[1] <- "Cluster_ID";
  colnames(p)[2] <- "Plot_ID";
  colnames(p)[3] <- "zone";
  colnames(p)[4] <- "X_map";
  colnames(p)[5] <- "Y_map"
  
  p$Cluster_ID <- rep(DT_Clusters$Cluster_ID, each = nplots)
  p$X          <- rep(DT_Clusters$X, each = nplots)
  p$Y          <- rep(DT_Clusters$Y, each = nplots)
  
  p$Plot_ID    <- rep(plot_list$No, times = nclus)
  p$X_map      <- rep(plot_list$X,  times = nclus)
  p$Y_map      <- rep(plot_list$Y,  times = nclus)
  p$zone       <- rep(DT_Clusters$zone, each = nplots)
  print("PLOT 1+")
  if ( map_projection==4326 ) {
      iii = 1
      for (ic in 1:nclus) {
        for (ip in 1:nplots) {
          if (plot_list$distance[ip] > 0) {
            target_coord <- geosphere::destPoint(c(p$X[iii], p$Y[iii]), plot_list$azimuth[ip], plot_list$distance[ip] )
            print(str(target_coord))
            p$X_map[iii] <- target_coord[1] 
            p$Y_map[iii] <- target_coord[2]
          } else {
            p$X_map[iii] <- p$X[iii] 
            p$Y_map[iii] <- p$Y[iii]
          }
          iii = iii + 1
        }
      }
  } else {
    clusters_X   <- rep(DT_Clusters$X_map, each = nplots)
    clusters_Y   <- rep(DT_Clusters$Y_map, each = nplots)
    
    p$X_map      <- p$X_map + clusters_X
    p$Y_map      <- p$Y_map + clusters_Y
  }
  p$X <- NULL
  p$Y <- NULL
  rm(clusters_X); rm(clusters_Y)
  
  print("PLOT 1++")
  # Get longlat coordinates for plot points --------
  plotdata <- split(p, p$zone)
  
  print("PLOT 2")
  
  for (i in 1:length(plotdata)) {
    
    df <- data.frame(plotdata[[i]])

    if (map_projection == 4326) {
      zone_str  <- as.character(names(plotdata)[i])
      zone      <- as.integer(substr(zone_str, 0, nchar(zone_str)-1))
      hemispere <- substr(zone_str, nchar(zone_str), nchar(zone_str))
            
      projcrs <- ifelse(hemispere=='S', 32700+zone, 32600+zone)
                         
      data <- st_as_sf(x = df,                         
                       coords = c("X_map", "Y_map"),
                       crs    = paste0("EPSG:", projcrs),
                       remove = TRUE)
      
      rm(hemispere); rm(zone_str)
    } else {
      data <- st_as_sf( x = df,                         
                       coords = c("X_map", "Y_map"),
                       crs    = paste0("EPSG:", map_projection),
                       remove = TRUE)
    }
 
     rm(df)

    if (map_projection != 4326 ) {
      res <- st_transform( data, 4326 )
    } else {
      res <- data
    }
    coord_list <- st_coordinates(res)
    
    res   <- st_drop_geometry(res)
    res$X <- coord_list[,1]
    res$Y <- coord_list[,2]
    
    rm(coord_list)
    res$zone <- NULL

    if (i == 1) {
      resAll <- res
    } else {
      resAll <- rbind(resAll, res)
    }
  }
  
  print("PLOT 3")
  
  p <- merge(x = p, y = resAll, by = c("Cluster_ID", "Plot_ID"), all.x = TRUE)
  p$Cluster_ID <- as.integer(p$Cluster_ID)
  p$Plot_ID    <- as.integer(p$Plot_ID)

  # order by Cluster_ID, Plot_ID
  p <- p[with(p, order(Cluster_ID, Plot_ID)),]

  if (map_projection==4326 & nplots==1 & mean(plot_list$X)==0 & mean(plot_list$Y)==0) {
    p$X_map = round(p$X_map, deci_x)
    p$X     = round(p$X    , deci_x)
    p$Y_map = round(p$Y_map, deci_y)
    p$Y     = round(p$Y    , deci_y)
  }
  
  p <- st_as_sf( x = p,                         
                 coords = c("X", "Y"),
                 crs    = "EPSG:4326",
                 remove = FALSE)

  rm(res); rm(resAll); rm(zone); rm(i); rm(data)
  print("PLOT 4")
  return(p)
}


ui.R

library('shinydashboard')
library('leaflet')
library('shinyjs')

ui <- function(request) {
  
  header <- dashboardHeader(
    title = "Cluster and Plot Grid Design Tool",
    titleWidth = 350
  )

  body <- dashboardBody(

   navbarPage("",
    tabPanel("Application",
     sidebarPanel(
                 
##    fluidRow(
      shinyjs::useShinyjs(),
##      column(width = 3,
          shinydashboard::box(width = NULL, status = "warning",
          h4("Actions"),
          actionButton(inputId = "run", label = "RUN"),
#          bookmarkButton(),
          downloadButton('downloadDataCSV'  , label = 'Download CSVs'),
          downloadButton('downloadDataShape', label = 'Download Shapes')
          
#            checkboxInput("drop_rowcol", "Export subset of rows/columns", FALSE),
#            numericInput("dropfactor", "Keep every Nth row/column", 1, min = 1, max = 100, step = 1,)
        ),

        # Map file parameters
          shinydashboard::box(width = NULL, status = "warning",
          h4("Upload map (optional)"),
          p(tags$small("Choose a polygon shapefile if you want to get the extent from a map")),

          # Input: Select a file ----
          # see https://gist.github.com/RCura/9587685
          fileInput(inputId="shpFile", label="Select shp, shx, dbf and prj", multiple=TRUE, accept=c(".shp",".shx",".dbf",".prj")),
        
          conditionalPanel(
            condition = "output.fileUploaded == true",
            checkboxInput("use_map_projection", "Use map projection", TRUE),
            checkboxInput("use_clipping_map", "Use map for clipping", TRUE)
          ),

          conditionalPanel(
            condition = "output.fileUploaded == true && input.use_clipping_map == false",
            checkboxInput("use_map_extent", "Get area corner coordinates from map extent", TRUE)
          ),
          
          textOutput("map_proj1")
        ),
        
        # Cluster parameters
          shinydashboard::box(width = NULL, status = "warning",
          h4("Cluster parameters"),

          conditionalPanel(condition = "(output.fileUploaded == true && input.use_map_projection == false) || output.fileUploaded == false", textInput("projection", "Projection", "EPSG:4326")),
          splitLayout(
            numericInput("minx", "Min X (West)", -170),
            numericInput("maxx", "Max X (East)",  170)),
          splitLayout(
            numericInput("miny", "Min Y (South)", -80),
            numericInput("maxy", "Max Y (North)",  80)),
          splitLayout(
            numericInput("cluster_x_distance", "X distance between clusters", 10),
            numericInput("cluster_y_distance", "Y distance between clusters", 10)),
          tags$small("Distances in meters, or in case of longlat as degrees. ",
                     tags$br(),tags$strong("Note:"),
                       "if using map extent, these must in the same units as map distance units"
          ),
          hr(),
           
          numericInput("first_cluster_id", "1st cluster number in the grid", 1),
          checkboxInput("cluster_start_random", "Randomize first cluster coordinates", FALSE),
#          checkboxInput("cluster_nested", "Create 3 nested grids for Collect Earth survey", FALSE),
          splitLayout(
            checkboxInput("cluster_hexagonal", "Hexagonal grid", FALSE),
            checkboxInput("cluster_hexagonal_dist", "Equal cluster distance (=X)", FALSE)
          ),
          conditionalPanel(
            condition = "!input.cluster_start_random",
            hr(),
            h5("1st cluster ref. point distance from left-upper corner of the area"),
            splitLayout(
              numericInput("cluster_x", "X distance", 0),
              numericInput("cluster_y", "Y distance", 0))
        )),
        hr(),
        # Plot parameters
          shinydashboard::box(width = NULL, status = "warning",
          h4("Plot parameters"),
          p("Plot location in relation to the cluster's reference point"),
          p(tags$small("Give ID, X and Y coordinates")),
          tags$div(id = "plots"),
          actionButton("insertPlotBtn", "Add plot"),
          hr(),
  
          #textInput("plot_list_n", "Plot list ids", "1, 2, 3, 4"),
          #textInput("plot_list_x", "Plot list X coordinates", "250, 250, 750, 750"),
          #textInput("plot_list_y", "Plot list Y coordinates", "250, 750, 750, 250"),
          selectInput("plot_type", "Plot type",
                      c("Circular/cocentric/relascope" = "C",
                        "Rectangular/square" = "R",
                        "NFMA" = "NFMA",
                        "Custom" = "Z")),
          conditionalPanel(
            condition = "input.plot_type != 'NFMA' & input.plot_type != 'Z'",
            selectInput("plot_reference_point", "Plot reference point location",
                        c("Center" = "C",
                          "NW" = "NW",
                          "SW" = "SW"))
          ),
          conditionalPanel(
            condition = "input.plot_type != 'Z'",
            numericInput("plot_dim_1", "Length: Radius or X axis (long.), in meters", 20)
          ),
          conditionalPanel(
            condition = "input.plot_type != 'C' & input.plot_type != 'Z'",
            numericInput("plot_dim_2", "Length: Y axis (lat.), in meters", 20)
          ),
          conditionalPanel(
            condition = "input.plot_type == 'Z'",
            wellPanel(
              p("Corner coordinates"),
              p(tags$small("Do not repeat the first coordinate")),
              textInput("plot_corner_x", "Plot corners' X coordinates", "0, 10, 10, -5"),
              textInput("plot_corner_y", "Plot corners' Y coordinates", "0, 0, 10, 10")
            )
          )
        )
      ),
mainPanel(
#      column(
#        width = 9,
        shinydashboard::box(width = NULL,
            solidHeader = TRUE,
            leafletOutput("plotmap", height = 800)
        ),
        
        shinydashboard::box(width = NULL, status = "warning",
          h4("Plot statistics"),
          tags$br(),
          splitLayout(
            h4(textOutput("number_clusters")),
            h4(textOutput("number_plots"))
          ),
          hr(),
          conditionalPanel(condition = "output.fileUploaded == true",
            splitLayout(
              numericInput("target_clusters", "Targeted number of clusters (optional)", 100),
              actionButton(inputId = "target_dist_button", label = "Estimate"),
              disabled(textInput("target_distance","Target distance"," ")),
              actionButton(inputId = "applydistance", label = "Apply distance")
            )
          )
        )
##            htmlOutput("html_link")
        )
      ),
 tabPanel("About", 
         includeMarkdown("www/releases.md")
         
 )
))
  
  dashboardPage(
    skin='green',
    header,
    dashboardSidebar(disable = TRUE),
    body
  )
  
}
